### PR TITLE
Saving Experiment Metadata in MC

### PR DESCRIPTION
**Pull request by lciorciaro**
Changes proposed in this pull request:
- Added an argument to MC.run() to pass a dictionary containing experimental metadata that will be saved to the hdf5 file of the experiment.

Note that #291 should be merged first, as some new functions from that PR are used here.

### CODE REVIEW

**Comment by lciorciaro** [comment]
The relevant file to look at is measurement_control.py. The other changes are merged from #291 .

**Comment by lciorciaro** [comment]
@AdriaanRol I wrote a simple test for the new feature.

File: pycqed/measurement/measurement_control.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def __init__(self, name: str,
     # Functions used to control the measurements #
     ##############################################
 
-    def run(self, name=None, mode='1D', **kw):
+    def run(self, name=None, exp_metadata=None, mode='1D', **kw):
         '''
```
Comment (AdriaanRol) [inline pycqed/measurement/measurement_control.py]: Could you add a docstring and type spec for the exp_metadata argument (and maybe also for the other input arguments while you are at it?)

File: pycqed/tests/test_hdf5_datasaving_loading.py [hunk 2]
```diff
@@ -83,6 +83,7 @@ def test_writing_and_reading_dicts_to_hdf5(self):
             'dataset2': np.array([[2, 3, 4, 5],
```
Comment (AdriaanRol) [inline pycqed/tests/test_hdf5_datasaving_loading.py]: Could you undo changes to this file?

File: pycqed/measurement/hdf5_data.py [hunk 3]
```diff
@@ -128,6 +128,7 @@ def encode_to_utf8(s):
     if isinstance(s, str):
```
Comment (AdriaanRol) [inline pycqed/measurement/hdf5_data.py]: Could you undo changes to this file?

File: pycqed/tests/test_MeasurementControl.py [hunk 4]
```diff
@@ -403,3 +405,31 @@ def test_data_resolution(self):
...
+        self.MC.set_detector_function(det.Dummy_Detector_Soft())
+        self.MC.run('test_exp_metadata', exp_metadata=metadata_dict)
+        a = ma.MeasurementAnalysis(label='test_exp_metadata', auto=False)
+
+        a_tools.datadir = old_a_tools_datadir
+
+        loaded_dict = read_dict_from_hdf5(
+            {}, a.data_file['Experimental Data']['Experimental Metadata'])
+
+        np.testing.assert_equal(metadata_dict, loaded_dict)
```
Comment (AdriaanRol) [inline pycqed/tests/test_MeasurementControl.py]: I'm not sure if np.testing works well for complex dictionaries. I always have trouble with testing equality of dicts (i.e., it returns True to soon). However, I have mostly used the default pytest assertEqual. Maybe this works exactly as expected. Let me know, then it should be OK.

File: pycqed/measurement/measurement_control.py [hunk 5]
```diff
@@ -839,6 +841,30 @@ def save_MC_metadata(self, data_object=None, *args):
         set_grp.attrs['measurement_name'] = self.measurement_name
         set_grp.attrs['live_plot_enabled'] = self.live_plot_enabled()
 
+    @classmethod
+    def save_exp_metadata(self, metadata: dict, data_object):
+        '''
+        Saves experiment metadata to the data file.
```
Comment (AdriaanRol) [inline pycqed/measurement/measurement_control.py]: Maybe add a comment as to what the location in the data object is it get's saved to.

**Comment by lciorciaro** [comment]
@AdriaanRol I implemented the changes you requested. As for the np.testing issue, this is from the documentation:

> Given two objects (scalars, lists, tuples, dictionaries or numpy arrays), check that all elements of these objects are equal. An exception is raised at the first conflicting values.

So it is supposed to work with dictionaries. I tried a few simple test cases by hand and it returns very detailed information about the mismatch.

**Pull request merged**